keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+
+#### Avoid double writing of logs using klog and logr APIs
+
+Instead of leaving the existing references to klog (or actually a klog wrapper) for an extended period of time until the legacy output format is completely removed from the Kubernetes and having to do a double writing to both klog and logr APIs for all existing logs we could consider two alternative approaches:
+* Use logr-based API for all existing and new logs.
+* Introduce intermediate API which would be a superset of the logr API and could produce exactly the same logs as the existing klog API.
+
+The main problem with the first approach is the fact that it makes very hard to produce log lines which would be exactly the same as the existing klog-based logs. This could be done for instance by adding special fields in logr-based logs called `legacy-message`, `legacy-level` and `legacy-source` which could be lazy evaluated and which could be used only by the logr legacy format.
+
+This however would make accidental breaking of the legacy output format compatibility much easier and would limit the flexibility when it comes to changes which can be introduced in the `json`/`dev` output format logs as it would be for instance harder to move API calls which produce those.
+
+The second approach is in fact very similar to the approach described in the previous section with the klog API enhancements. Apart from the problems described there it has the same problem with the flexibility which exists for using logr-based API for existing logs. The other problem is that it introduces yet another logging API which might cause some confusion and will make working on multiple versions of the Kubernetes at the same time harder.
+
+The requirement of being able to produce logs in the same format is important for two reasons:
+* Although logs have never been considered an API for the Kubernetes and Kubernetes does not provide any guarantees when it comes to logs stability there are still some Kubernetes users which rely on the existing logs for debugging, monitoring or alerting purposes. Adding an option to get access to logs in the old format will provide safer and more reliable path for upgrade for those users.
+* It will be hard to come up with a stable set of log messages and fields in the new json/dev format in a single Kubernetes release. Therefore for instance log schema might have a breaking changes before new logs API goes to GA. Depending on the logs storage solution being used this might impose some problems from an operational point of view. To address this issue Kubernetes users which are concerned about this will have an option to stick to the legacy logs format until new format is completely stable.
```

Comment:
That is potentially a bigger issue - I agree.
But instead of double-writing, I can also imagine a temporary step where we will have:
```
logger.Infof("whatever we want to product now").OldFormat("whatever klog is producing")
```
and the library was dropping one of those depending on the setting (in legacy mode, it would simply call klog internally)

And at the implementation level we can guarantee no double-writing.

Why is that a bad idea?

@serathius @44past4

Reply:
To have a full compatibility for logs content for legacy log format and not to limit ourselves when it comes to the functionalities provided by logr the new temporary API would need to contain most methods from both libraries with little overlap between them. Therefore the benefits coming from introduction of such API would be limited and the costs related for instance to the need of learning of new API would be quite high.